File: lib/roundtable/prompt/roles.ex
defmodule Roundtable.Prompt.Roles do
  @moduledoc "Loads role prompt files with project-local → global fallback."

  @spec load_role_prompt(String.t(), String.t(), String.t() | nil) :: String.t()
  def load_role_prompt(role_name, global_dir, project_dir \\ nil) do
    filename = "#{role_name}.txt"

    if project_dir do
      case File.read(Path.join(project_dir, filename)) do
        {:ok, content} ->
          content

        {:error, :enoent} ->
          load_from_global(role_name, global_dir, project_dir)

        {:error, reason} ->
          raise "Cannot read role file #{role_name} in project dir: #{inspect(reason)}"
      end
    else
      load_from_global(role_name, global_dir, nil)
    end
  end

  defp load_from_global(role_name, global_dir, project_dir) do
    filename = "#{role_name}.txt"

    case File.read(Path.join(global_dir, filename)) do
      {:ok, content} ->
        content

      {:error, :enoent} ->
        searched = if project_dir, do: project_dir, else: "none"
        raise "Role prompt not found: #{role_name} (searched #{searched}, #{global_dir})"

      {:error, reason} ->
        raise "Cannot read role file #{role_name} in global dir: #{inspect(reason)}"
    end
  end
end

File: internal/roundtable/roles.go
package roundtable

import (
	"embed"
	"fmt"
	"os"
	"path/filepath"
)

//go:embed roles/*.txt
var embeddedRoles embed.FS

// LoadRolePrompt loads a role prompt by name. Fallback semantics match Elixir:
//
//  1. projectDir/<name>.txt  (if projectDir is non-empty)
//  2. globalDir/<name>.txt   (if globalDir is non-empty)
//  3. If BOTH dirs were set and the file wasn't found → error (matching Elixir)
//  4. If globalDir was empty/unset (standalone binary mode) → embedded defaults
//  5. If embedded doesn't have it either → error
//
// Returns the file contents as a string. Returns an error if the role
// is not found or if a file exists but cannot be read.
func LoadRolePrompt(roleName, globalDir, projectDir string) (string, error) {
	filename := roleName + ".txt"

	// Level 1: project directory
	if projectDir != "" {
		content, err := os.ReadFile(filepath.Join(projectDir, filename))
		if err == nil {
			return string(content), nil
		}
		if !os.IsNotExist(err) {
			return "", fmt.Errorf("cannot read role file %s in project dir: %w", roleName, err)
		}
	}

	// Level 2: global directory
	if globalDir != "" {
		content, err := os.ReadFile(filepath.Join(globalDir, filename))
		if err == nil {
			return string(content), nil
		}
		if !os.IsNotExist(err) {
			return "", fmt.Errorf("cannot read role file %s in global dir: %w", roleName, err)
		}
	}

	// If any directory was configured, disk lookup is authoritative — do NOT
	// silently fall back to embedded defaults. This matches Elixir, which
	// raises when role is not found in either directory.
	if globalDir != "" || projectDir != "" {
		searched := []string{}
		if projectDir != "" {
			searched = append(searched, projectDir)
		}
		if globalDir != "" {
			searched = append(searched, globalDir)
		}
		return "", fmt.Errorf("role prompt not found: %s (searched %v)", roleName, searched)
	}

	// Level 3: embedded defaults — only when globalDir is empty/unset
	// (standalone binary mode, no roles directory configured).
	content, err := embeddedRoles.ReadFile("roles/" + filename)
	if err == nil {
		return string(content), nil
	}

	return "", fmt.Errorf("role prompt not found: %s (searched [embedded])", roleName)
}
